### PR TITLE
Added cascade ensemble PoC

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

*Description of changes:*

Added cascade ensemble PoC.

The goal of a cascade ensemble is to expand the Accuracy - Inference Latency Pareto Frontier by finding a middle ground between single models and full ensembles that gets the majority of the accuracy improvement of ensembles combined with the majority of the inference speed of single models.

Currently, usage / existence of cascade ensemble support is not user-facing. It will be documented once it has been further tested and optimized. Currently the cascade threshold is fixed at 0.9 and the cascade must be manually defined. Next steps are to automatically define the cascade and the thresholds.

The below example script shows how to use cascades to get 80% of the log loss improvement (97% of the accuracy improvement) of the ensemble over a single model with only 35% the inference cost, thus expanding the Pareto Frontier, without any tuning.

Example Script:

```
from autogluon.tabular import TabularPredictor, TabularDataset


if __name__ == '__main__':
    path_prefix = 'https://autogluon.s3.amazonaws.com/datasets/CoverTypeMulticlassClassification/'
    path_train = path_prefix + 'train_data.csv'
    path_test = path_prefix + 'test_data.csv'

    label = 'Cover_Type'
    sample = 50000  # Number of rows to use to train
    train_data = TabularDataset(path_train)

    if sample is not None and (sample < len(train_data)):
        train_data = train_data.sample(n=sample, random_state=0).reset_index(drop=True)

    test_data = TabularDataset(path_test)

    fit_kwargs = dict(
        train_data=train_data,
        hyperparameters={
            'NN_TORCH': {},
            'GBM': {},
            'KNN': {},
        },
    )

    predictor = TabularPredictor(
        label=label,
        eval_metric='log_loss',
    )

    predictor.fit(**fit_kwargs)

    predictor.persist_models('all')

    leaderboard = predictor.leaderboard(test_data)

    import time

    ts = time.time()
    pred_proba = predictor.predict_proba(test_data, model='WeightedEnsemble_L2')
    te = time.time()

    print('--------')
    print('Normal: ')
    print(predictor.evaluate_predictions(y_true=test_data[label], y_pred=pred_proba, silent=True))
    print(f'{te - ts}s')
    print('--------')

    ts = time.time()
    pred_proba = predictor.predict_proba(test_data, model=['NeuralNetTorch', 'WeightedEnsemble_L2'])
    te = time.time()

    print('--------')
    print('Cascade:')
    print(predictor.evaluate_predictions(y_true=test_data[label], y_pred=pred_proba, silent=True))
    print(f'{te - ts}s')
    print('--------')

```

Out:

```
                 model  score_test  score_val  pred_time_test  pred_time_val     fit_time  pred_time_test_marginal  pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  fit_order
0  WeightedEnsemble_L2   -0.234278  -0.260248       14.950171       0.715236  1270.274406                 0.048993                0.001165           0.486316            2       True          4
1             LightGBM   -0.295047  -0.324501       13.393521       0.607209    11.692870                13.393521                0.607209          11.692870            1       True          2
2       NeuralNetTorch   -0.297718  -0.327164        0.851722       0.072343  1258.052863                 0.851722                0.072343        1258.052863            1       True          3
3           KNeighbors   -0.496720  -0.456807        0.655935       0.034519     0.042357                 0.655935                0.034519           0.042357            1       True          1


--------
Normal: 
{'log_loss': -0.23427786256734967, 'accuracy': 0.9106477457552731, 'balanced_accuracy': 0.8266616778634864, 'mcc': 0.8558472797101921}
15.58s
--------
--------
Cascade:
{'log_loss': -0.25118394508483005, 'accuracy': 0.9099506897412287, 'balanced_accuracy': 0.8307863237257622, 'mcc': 0.8547748671813279}
6.10s
--------
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: core/src/autogluon/core/trainer/abstract_trainer.py [hunk 1]
```diff
@@ -573,46 +584,177 @@ def score_with_y_pred_proba(self, y, y_pred_proba, weights=None) -> float:
         return compute_weighted_metric(y, y_pred, self.eval_metric, weights, weight_evaluation=self.weight_evaluation,
                                        quantile_levels=self.quantile_levels)
 
+    # TODO: Slow if large ensemble with many models, could cache output result to speed up cascades during inference
+    def _construct_model_pred_order(self, models: List[str]) -> List[str]:
```
Comment (gradientsky) [inline core/src/autogluon/core/trainer/abstract_trainer.py:588]: tests

File: core/src/autogluon/core/trainer/abstract_trainer.py [hunk 2]
```diff
@@ -573,46 +584,177 @@ def score_with_y_pred_proba(self, y, y_pred_proba, weights=None) -> float:
...
+                continue
+            min_models_set = set(self.get_minimum_model_set(model))
+            models_to_load = list(min_models_set.difference(model_set))
+            subgraph = nx.subgraph(self.model_graph, models_to_load)
+            model_pred_order = list(nx.lexicographical_topological_sort(subgraph))
+            model_order += [m for m in model_pred_order if m not in model_set]
+            model_set = set(model_order)
+        return model_order
+
+    def _construct_model_pred_order_with_pred_dict(self, models: List[str], models_to_ignore: List[str] = None) -> List[str]:
```
Comment (gradientsky) [inline core/src/autogluon/core/trainer/abstract_trainer.py:618]: tests

File: core/src/autogluon/core/trainer/abstract_trainer.py [hunk 3]
```diff
@@ -638,6 +792,45 @@ def get_model_pred_proba_dict(self, X, models, model_pred_proba_dict=None, model
...
+                # If model is part of cascade, keep the predictions that are confident and don't predict on these rows with further models.
+                if model_name in models and model_name != models[-1]:
+                    pred_proba = model_pred_proba_dict[model_name]
+                    # Calculate confident predictions based on cascade threshold
+                    # TODO: Support more sophisticated methods of calculating whether to keep a prediction
+                    # TODO: Support per-model confidence specification
+                    if self.problem_type == BINARY:
+                        confident = (pred_proba >= cascade_threshold) | (pred_proba <= (1-cascade_threshold))
+                    elif self.problem_type == MULTICLASS:
+                        confident = (pred_proba >= cascade_threshold).any(axis=1)
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply core/src/autogluon/core/trainer/abstract_trainer.py:813]: For multiclass we can also consider “high probability” based on relative weights: like if we have 51 class and 50 values are 0.01 and one 0.5 - that’s a high-confidence prediction, but if all they are very close, then it’s not

[the remark this answers is not in the document]
Comment (Innixma) [reply core/src/autogluon/core/trainer/abstract_trainer.py:813]: This can be future work, I'd prefer to keep it simple currently.

**Comment by szha** [comment]
Job PR-1682-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1682/4/index.html

**Comment by Innixma** [comment]
Closing, refer to updated version: #1929

**Pull request closed without merging**